libkbfs/delayed_cancellation.go

```diff
@@ -0,0 +1,205 @@
+// also makes this change replayable by NewContextWithReplayFrom. When
+// replayed, the resulting context is replayable as well.
+//
+// It is important that all WithValue-isu mutations on ctx is done "replayably"
+// (with NewContextReplayable) if any delayed cancellation is used, e.g.
+// through EnterCriticalWithTimeout,
+func NewContextReplayable(
+	ctx context.Context, change CtxReplayFunc) context.Context {
+	var replay CtxReplayFunc
+	replay = func(ctx context.Context) context.Context {
+		ctx = change(ctx)
+		// _ is necessary to avoid panic if nil
+		replays, _ := ctx.Value(CtxReplayKey).([]CtxReplayFunc)
+		replays = append(replays, replay)
+		return context.WithValue(ctx, CtxReplayKey, replays)
+	}
+	return replay(ctx)
+}
+
+// NewContextWithReplayFrom constructs a new context out of ctx by calling all
+// attached replay functions. This disconnects any existing context.CancelFunc.
+func NewContextWithReplayFrom(ctx context.Context) (context.Context, error) {
+	if replays, ok := ctx.Value(CtxReplayKey).([]CtxReplayFunc); ok {
+		newCtx := context.Background()
+		for _, replay := range replays {
+			newCtx = replay(newCtx)
+		}
+		return newCtx, nil
+	}
+	return nil, CtxNotReplayableError{}
+}
+
+type criticalAwareness struct {
+	*sync.Cond
+	isCritical bool
+	timer      *time.Timer
+	canceled   bool
+}
+
+func newCriticalAwareness() *criticalAwareness {
+	return &criticalAwareness{
+		Cond:       sync.NewCond(&sync.Mutex{}),
+		isCritical: false,
+		timer:      nil,
+		canceled:   false,
+	}
+}
+
+// NewContextWithCriticalAwareness creates a new context out of ctx. All replay
+// functions attached to ctx are run on the new new context. In addition, the
+// new context is made "critical aware". That is, it disconnects the cancelFunc
+// from ctx, and watch for the cancellation. When cancellation happens, it
+// checks if the associated context is in critical state, and wait until it
+// exits the critical state before cancelling the new context. This provides a
+// hacky way to allow finer control over cancellation.
+//
+// Note that, it's important to call context.WithCancel (or its friends) before
+// this function if those cancellations need to be controllable (critical
+// aware). Otherwise, the new cancelFunc is inheriently NOT "critical aware".
+func NewContextWithCriticalAwareness(
+	ctx context.Context) (newCtx context.Context, err error) {
+	v := ctx.Value(CtxCriticalAwarenessKey)
+	if v != nil {
+		if _, ok := v.(*criticalAwareness); ok {
+			return nil, ContextAlreadyHasCriticalAwarenessError{}
+		}
+		return nil, NoCriticalAwarenessError{}
+	}
+
+	if newCtx, err = NewContextWithReplayFrom(ctx); err != nil {
+		return nil, err
+	}
+	c := newCriticalAwareness()
+	newCtx = NewContextReplayable(newCtx,
+		func(ctx context.Context) context.Context {
+			return context.WithValue(ctx, CtxCriticalAwarenessKey, c)
+		})
+	newCtx, cancel := context.WithCancel(newCtx)
+	go func() {
+		<-ctx.Done()
```

Comment:
If `ctx` didn't have a canceled function, or it's never called, then this will leak goroutines.  Maybe you could save `cancel` in the struct, call it in `ExitCritical`, and select on it here?